cmd/envoy/main.go

```diff
@@ -781,3 +788,34 @@ func valueForColumn(table, column string) interface{} {
 
 	panic(fmt.Errorf("unknown type for %s.%s", table, column))
 }
+
+func UpdateCounterpartyAndContacts(ctx context.Context, modelCounterparty *models.Counterparty, apiCounterparty *api.Counterparty) error {
+	// update the counterparty
+	if err := db.UpdateCounterparty(ctx, modelCounterparty); err != nil {
+		return err
+	}
+
+	// update or create contacts (not performed in `UpdateCounterparty()`)
+	for _, apiContact := range apiCounterparty.Contacts {
+		// convert the api.Contact to a model.Contact
+		var modelContact *models.Contact
+		var err error
+		if modelContact, err = apiContact.Model(modelCounterparty); err != nil {
+			log.Warn().Err(err).Msg(fmt.Sprintf("Error when converting Contact to a model: '%s'", apiContact.Name))
+			continue
+		}
+
+		// create the contact, updating any already-existing contacts
+		if err = db.CreateContact(ctx, modelContact); err != nil {
+			if errors.Is(err, dberr.ErrAlreadyExists) {
+				//TODO: get the contact's ID and update the contact (this improvement has a ticket)
+				continue
+			} else {
+				log.Warn().Err(err).Msg(fmt.Sprintf("Error when creating Contact: '%s' (CounterpartyID: '%s')", modelContact.Name, modelCounterparty.ID))
+				continue
+			}
+		}
```

Comment:
I'm going to use `modelContact.SetContacts` here instead when I push my recommendations; I think that will make the single transaction model a bit easier to read and handle.